tuned/plugins/plugin_irq.py

```diff
@@ -0,0 +1,269 @@
+						% (mode, instance.name))
+			instance._mode = "set"
+
+	def _instance_cleanup(self, instance):
+		pass
+
+	def _instance_apply_static(self, instance):
+		log.debug("Applying IRQ affinities (%s)" % instance.name)
+		super(IrqPlugin, self)._instance_apply_static(instance)
+
+	def _instance_unapply_static(self, instance, rollback):
+		log.debug("Unapplying IRQ affinities (%s)" % instance.name)
+		super(IrqPlugin, self)._instance_unapply_static(instance, rollback)
+
+	def _instance_verify_static(self, instance, ignore_missing, devices):
+		log.debug("Verifying IRQ affinities (%s)" % instance.name)
+		return super(IrqPlugin, self)._instance_verify_static(instance, ignore_missing, devices)
+
+	#
+	# "low-level" methods to get/set irq affinities
+	#
+	def _get_irq_affinity(self, irq):
+		"""Get current IRQ affinity from the kernel
+
+		Args:
+			irq (str): IRQ number (as string) or "DEFAULT"
+
+		Returns:
+			affinity (set): set of all CPUs that belong to the IRQ affinity mask
+		"""
+		filename = "/proc/irq/default_smp_affinity" if irq == "DEFAULT" else "/proc/irq/%s/smp_affinity" % irq
+		with open(filename, "r") as f:
+			affinity_hex = f.readline().strip()
+		return set(self._cmd.hex2cpulist(affinity_hex))
+
+	def _set_irq_affinity(self, irq, affinity, restoring):
+		"""Set IRQ affinity in the kernel
+
+		Args:
+			irq (str): IRQ number (as string) or "DEFAULT"
+			affinity (set): affinity mask as set of CPUs
+			restoring (bool): are we rolling back a previous change?
+
+		Returns:
+			status (int):  0 on success, -2 if changing the affinity is not
+				supported, -1 if some other error occurs
+		"""
+		try:
+			affinity_hex = self._cmd.cpulist2hex(list(affinity))
+			log.debug("Setting SMP affinity of IRQ %s to '%s'"
+					% (irq, affinity_hex))
+			filename = "/proc/irq/default_smp_affinity" if irq == "DEFAULT" else "/proc/irq/%s/smp_affinity" % irq
+			with open(filename, "w") as f:
+				f.write(affinity_hex)
+			return 0
+		except (OSError, IOError) as e:
+			# EIO is returned by
+			# kernel/irq/proc.c:write_irq_affinity() if changing
+			# the affinity is not supported
+			# (at least on kernels 3.10 and 4.18)
+			if hasattr(e, "errno") and e.errno == errno.EIO \
+					and not restoring:
+				log.debug("Setting SMP affinity of IRQ %s is not supported"
+						% irq)
+				return -2
+			else:
+				log.error("Failed to set SMP affinity of IRQ %s to '%s': %s"
+						% (irq, affinity_hex, e))
+				return -1
+
+	#
+	# "high-level" methods: apply tuning while saving original affinities
+	#
+	def _apply_irq_affinity(self, irqinfo, affinity):
+		"""Apply IRQ affinity tuning
+
+		Args:
+			irqinfo (IrqInfo): IRQ that should be tuned
+			affinity (set): desired affinity
+		"""
+		instance = self._get_device_instance(irqinfo.device)
+		if instance is None:
+			return
+		original = self._get_irq_affinity(irqinfo.irq)
+		if instance._mode == "intersect":
+			affinity = affinity & original or affinity
```

Comment:
Maybe add comment here for better readability, e.g.:
```
# intersection of affinity and original, affinity if the intersection is empty
```

Reply:
done.